src/write_statsd.c

```diff
@@ -0,0 +1,476 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * Authors:
+ *   Stefano Pogliani <stefano at spogliani.net>
+ **/
+
+#define _BSD_SOURCE
+
+#include <limits.h>
+#include <stdlib.h>
+
+#include <netdb.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include "collectd.h"
+#include "common.h"
+#include "plugin.h"
+
+
+/*** Compile-time constants. ***/
+#define WS_DOUBLE_FORMAT  "%f"
+#define VALUE_STR_LEN     20
+#define WRITE_STATSD_NAME "write_statsd"
+
+#define FREE_NOT_NULL(var) \
+  if (var != NULL) {       \
+    free(var);             \
+    var = NULL;            \
+  }
+
+
+/*** Type declarations. ***/
+struct write_statsd_config_s {
+  char* host;
+  int   port;
+  char* postfix;
+  char* prefix;
+  _Bool silence_type_warnings;
+  _Bool always_append_ds;
+
+  /* Internally derived options. */
+  char*  event_line_format;
+  size_t event_line_base_len;
+  /* The short line format is used when the DS name is not included. */
+  char*  short_event_line_format;
+  size_t short_event_line_base_len;
+};
+typedef struct write_statsd_config_s write_statsd_config_t;
+
+struct write_statsd_link_s {
+  struct sockaddr_in server;
+  int    sock;
+};
+typedef struct write_statsd_link_s write_statsd_link_t;
+
+
+/*** Module variables. ***/
+static char* DS_TYPE_TO_STATSD[] = {
+  "c",  /* DS_TYPE_COUNTER  */
+  "g",  /* DS_TYPE_GAUGE    */
+  "g",  /* DS_TYPE_DERIVE   */
+  NULL  /* DS_TYPE_ABSOLUTE */
+};
+
+/*
+ * The configuration is stored in a global instance and must not
+ * be modified outside of load_config or initialise.
+ */
+static write_statsd_config_t configuration = {
+  NULL, 8125, NULL, NULL, 0, 0,
+  NULL, 0, NULL, 0
+};
+
+
+/*** Module functions. ***/
+static void* allocate(size_t size) {
+  void* mem = (char*)malloc(size);
+  if (mem == NULL) {
+    ERROR("write_statsd plugin: not enough memory.");
+    return NULL;
+  }
+  return mem;
+}
+
+static char* ds_value_to_string(int type, value_t value) {
+  char* result = allocate(VALUE_STR_LEN);
+  if (result == NULL) {
+    return NULL;
+  }
+
+  switch(type) {
+    case DS_TYPE_COUNTER:
+      ssnprintf(result, VALUE_STR_LEN, "%lld", value.counter);
+      break;
+
+    case DS_TYPE_GAUGE:
+      ssnprintf(result, VALUE_STR_LEN, WS_DOUBLE_FORMAT, value.gauge);
+      break;
+
+    case DS_TYPE_DERIVE:
+      ssnprintf(result, VALUE_STR_LEN, "%"PRId64, value.derive);
+      break;
+
+    default:
+      free(result);
+      return NULL;
+  }
+
+  return result;
+}
+
+
+static write_statsd_link_t open_socket(void) {
+  write_statsd_link_t link  = {{0}, 0};
+  struct addrinfo* resolved = NULL;
+  int error = 0;
+
+  link.sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
```

Comment:
Please rewrite this in a protocol independent manner. Look for `getaddrinfo()` in `src/zookeeper.c` for an example of opening a client connection.